lib/madwf_ml.cpp

```diff
@@ -0,0 +1,347 @@
+        pushVerbosity(QUDA_SILENT);
+        base(ATphi, *forward_tmp);
+        popVerbosity();
+
+        // D' * A * T * phi
+        madwf_ml::transfer_5d_hh(theta, ATphi, P, true);
+
+        // ATx(ADphi, *phi, P);
+        madwf_ml::transfer_5d_hh(*forward_tmp, *phi, P, false);
+        pushVerbosity(QUDA_SILENT);
+        base(ADphi, *forward_tmp);
+        popVerbosity();
+
+        // T' * A * D * phi
+        madwf_ml::transfer_5d_hh(tmp, ADphi, device_param, true);
+        // theta
+        blas::axpy(1.0, theta, tmp);
+        if (tune_suppressor) { blas::axpy(pmu, *phi, tmp); }
+
+        ref(theta, tmp);
+
+        // lambda = D' * A * D * phi
+        madwf_ml::transfer_5d_hh(tmp, ADphi, P, true);
+
+        ref(lambda, tmp);
+
+        std::vector<ColorSpinorField *> lhs {&chi, &theta, &lambda};
+        std::vector<ColorSpinorField *> rhs {&chi, &theta, &lambda};
+        Complex dot[9];
+        blas::cDotProduct(dot, lhs, rhs);
+
+        a[0] += dot[0].real();
+        a[1] += -2.0 * dot[1].real();
+        a[2] += dot[4].real() + 2.0 * dot[2].real();
+        a[3] += -2.0 * dot[5].real();
+        a[4] += dot[8].real();
+      }
+
+      std::array<double, 4> coeffs = {4.0 * a[4], 3.0 * a[3], 2.0 * a[2], a[1]};
+      auto rs = cubic_formula(coeffs);
+
+      alpha = 0;
+      double root_min = poly4(a, 0);
+      for (auto r : rs) {
+        double eval = poly4(a, r);
+        if (root_min > eval) {
+          root_min = eval;
+          alpha = r;
+        }
+      }
+
+      axpby(device_param, 0.0f, device_param, -alpha, P);
+      if (tune_suppressor) { mu -= alpha * pmu; }
+
+      if (getVerbosity() >= QUDA_SUMMARIZE) {
+        printfQuda("grad min iter %05d: %04d chi2 = %8.4e, chi2 %% = %8.4e, alpha = %+8.4e, mu = %+8.4e\n", comm_rank(),
+                   iteration, chi2, chi2 / residual, alpha, mu);
+      }
+    }
+
+    trained = true;
+
+    if (getVerbosity() >= QUDA_VERBOSE) { printfQuda("Training finished ...\n"); }
+    count = 0;
+    for (const auto &phi : B) {
+      double ind_chi2 = cost(ref, base, chi, *phi);
+      double phi2 = blas::norm2(*phi);
+      if (getVerbosity() >= QUDA_VERBOSE) {
+        printfQuda("chi2 %03d %% = %8.4e, phi2 = %8.4e\n", count, ind_chi2 / phi2, phi2);
+      }
+      count++;
+    }
+    profile.TPSTOP(QUDA_PROFILE_TRAINING);
+
+    if (param.madwf_param_save) {
+      save_parameter(Ls, Ls_base);
+      comm_barrier();
+    }
+
+    // Destroy all dynamically allocated stuff.
+    for (auto &pB : B) { delete pB; }
+
+    commGlobalReductionPop();
+  }
+
+  void MadwfAcc::save_parameter(int Ls, int Ls_base) {
+    profile.TPSTART(QUDA_PROFILE_IO);
+    std::vector<transfer_float> host_param = device_param.to_host();
+
+    std::string save_param_path(param.madwf_param_outfile);
+    char cstring[512];
+    sprintf(cstring, "/madwf_trained_param_rank_%05d_ls_%02d_%02d_mu_%.3f.dat", comm_rank(), Ls, Ls_base, mu);
```

Comment:
This should also encode the decomposition (gridsize) to be safe. I could imagine someone copying these files from, say, Summit to Perlmutter, where you'd use a different decomposition.

Reply:
In principle the parameters here should be independent of the decomposition (it is purely a 5th dimension thing), so for the same lattice one should be able to use the same parameter for different decomposition. But in practice the decomposition could matter. I guess the correct thing to do is allowing user to add a customized string so they have the control over this?

Reply:
Ah, okay, so I misunderstood---I thought there were different parameters for different domains. In that case, I'm guessing only one rank should be responsible for load/save, and the `load` in particular should broadcast to other ranks.

Reply:
Otherwise I'm not sure why each rank needs to write out its own file

Reply:
Per our discussion, for now, we/I should change the code so that it only runs training on one rank, and at the end broadcasts the trained parameter to all ranks.

Reply:
Well thought about it for a while, only running training on rank zero is fine, but if it is not rank zero, it would conflict with the fact that only rank zero runs tuning for non-dslash kernels. I think for now we should still run training on all ranks, but only write out one parameter to the disk?

Reply:
https://github.com/lattice/quda/pull/1061/commits/60e2f6d2f499192c4eb9dd4812a22c8e76bfd6cd